src/nvim/eval.c

```diff
@@ -6175,9 +6175,10 @@ bool read_blob(FILE *const fd, blob_T *const blob)
 /// @param[in]  tv   Value to store as a string.
 /// @param[out] len  Length of the resulting string or -1 on error.
 /// @param[in]  endnl If true, the output will end in a newline (if a list).
+/// @param[in]  crlf  If true, list items will be joined with CRLF (if a list).
 /// @returns an allocated string if `tv` represents a VimL string, list, or
 ///          number; NULL otherwise.
-char *save_tv_as_string(typval_T *tv, ptrdiff_t *const len, bool endnl)
+char *save_tv_as_string(typval_T *tv, ptrdiff_t *const len, bool endnl, bool crlf)
```

Comment:
could change the `endnl` param to a `TriState` or `int` where:

```
0 = none
1 = LF
2 = CRLF
```

- And `assert(endnl <= 2)`.
- This eliminates the `crlf` param.
- Otherwise there's an implicit coupling between `endnl` and `crlf`. if crlf=true and endnl=false, that doesn't make sense.

Reply:
>     if crlf=true and endnl=false, that doesn't make sense.

But this condition may exist, can't it? In L6238, we're checking only for CRLF vs LF; the value of `endnl` doesn't enter here. What should be the value of the TriState var for this line when `endnl=false`?

Reply:
there's no reason for it to exist so we should prevent it

Reply:
Sadly, still confused.

There are two places where `endnl` and `crlf` are used.

```
    *len += (ptrdiff_t)strlen(tv_get_string(TV_LIST_ITEM_TV(li))) + (crlf ? 2 : 1);  // Ln6238
<snip>
  char *ret = xmalloc((size_t)(*len) + (endnl ? (crlf ? 2 : 1) : 0));                          // Ln6245
```

In the second case, I understand there are three possible values, so a tristate variable. But if `endnl` is 0, which _is_ the case with the two places this function is called from<sup>[1],[2]</sup>, how am I to know whether to use `CRLF` or `LF` in Ln6238?

I may be completely mistaking your point.


[1]: https://github.com/3N4N/neovim/blob/05d3625486a7/src/nvim/eval/funcs.c#L809
[2]: https://github.com/3N4N/neovim/blob/05d3625486a7/src/nvim/eval.c#L5691

Reply:
You are right. I should have looked closer. I thought that `endnl` was controlling if the lines were joined at all.